service/src/main/java/org/mskcc/smile/service/impl/TempoMessageHandlingServiceImpl.java

```diff
@@ -347,6 +358,66 @@ public void run() {
         }
     }
 
+    // TODO: move this function to TempoService and TempoServiceImpl? (except for the publish part)
+    // TODO: break this up into smaller functions?
+    private void publishTempoSamplesToCBioPortal(Set<String> tumorPrimaryIds) throws Exception {
+        // validate and build tempo samples to publish to cBioPortal
+        Set<TempoSample> validTempoSamples = new HashSet<>();
+        for (String primaryId : tumorPrimaryIds) {
+            try {
+                // confirm sample exists by primary id
+                SmileSample sample = sampleService.getDetailedSampleByInputId(primaryId);
+                if (sample == null) {
+                    LOG.error("Sample not found with primary ID: " + primaryId);
+                    continue;
+                }
+                // confirm tempo data exists by primary id
+                Tempo tempo = tempoService.getTempoDataBySamplePrimaryId(primaryId);
+                if (tempo == null) {
+                    LOG.error("Tempo data not found for sample with primary ID: " + primaryId);
+                    continue;
+                }
+                // validate access level and custodian information before building tempo sample
+                String accessLevel = tempo.getAccessLevel();
+                String custodianInfo = tempo.getCustodianInformation();
+                if (accessLevel == null || accessLevel.trim().isEmpty()) {
+                    LOG.error("Invalid access level for sample with primary ID: " + primaryId);
+                    continue;
+                }
+                if (custodianInfo == null || custodianInfo.trim().isEmpty()) {
+                    LOG.error("Invalid custodian information for sample with primary ID: " + primaryId);
+                    continue;
+                }
+                // build tempo sample object
+                TempoSample tempoSample = TempoSample.newBuilder()
+                    .setPrimaryId(primaryId)
+                    .setCmoSampleName(sample.getLatestSampleMetadata().getCmoSampleName())
+                    .setAccessLevel(accessLevel)
+                    .setCustodianInformation(custodianInfo)
+                    .build();
+                validTempoSamples.add(tempoSample);
```

Comment:
You can set up a new [constructor](https://www.w3schools.com/java/java_constructors.asp)

Reply:
oh actually I'm seeing now that this is a protobuf object? disregard my comment if that's the case

Reply:
Yep! This is the pattern to build a protobuf object per their docs.